interpreter/resolver.go

```diff
@@ -0,0 +1,564 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"reflect"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+
+	anypb "github.com/golang/protobuf/ptypes/any"
+	structpb "github.com/golang/protobuf/ptypes/struct"
+)
+
+// Resolver interface defines methods for resolving both absolute and relative attributes.
+//
+// One of the most expensive operations in CEL is object accessor reflection. When an Attribute
+// appears in the CEL expression, in most cases, the intent of the user is to select a leaf node
+// of a complex object which is of a simple primitive type. In these scenarios, the Resolver is
+// ideally situated to look at the Attribute path and find the most optimal way to retrieve the
+// data.
+//
+// For applications which use proto-based inputs or custom types with relatively static
+// environments are encouraged to hand-roll Resolver implementations suited to their strongly
+// typed objects.
+type Resolver interface {
+	// Resolve finds a top-level Attribute from the given Activation, returning the value
+	// if present.
+	//
+	// When the resolver cannot find the Attribute in the activation it must return nil, false.
```

Comment:
When the resolver cannot find _the Attribute's variable_ in the activation, it returns nil, false.  If one of the Attribute's path elements cannot be found, returns `e, true` where `e` is a `types.Err` value.